### PR TITLE
test(agent-policy): add OPA policy engine test suite

### PR DESCRIPTION
**Pull request by SantoshMadhukar-K**
Add a comprehensive OPA policy engine test suite for the `kata-agent-policy` crate, covering all major agent RPC endpoints via the Regorus evaluation engine.

## Changes

- **`src/agent/policy/src/policy.rs`** — new `#[cfg(test)]` tests using real `protocols::agent` request structs from `rpc.rs` serialised via `serde_json`
- **`src/agent/policy/Cargo.toml`** — add `[dev-dependencies]`: `rstest`, `tokio` (test features), `protobuf`
- **`Cargo.lock`** — updated

## Test coverage

- `CopyFileRequest` — file_type/path matrix (Regular, Directory, Symlink) with traversal and symlink-target validation
- `ExecProcessRequest` — Args, SELinux label, and AppArmor profile enforcement
- `CreateSandboxRequest` — guest_hook_path, kernel_modules, sandbox_pidns, and regex hostname matching
- `UpdateRoutesRequest` — forbidden source and device filtering
- Bundled policy smoke-tests (`allow-all`, `allow-all-except-exec-process`, `allow-set-policy`)
- Engine lifecycle — `set_policy` atomicity, `AllowRequestsFailingPolicy` bypass, `initialize()` success and missing-file error paths
- `print()` trace observability on both allow and deny paths

```
cargo test -p kata-agent-policy

warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/ksmadhukar/upstream-repos/kata-containers/src/agent/Cargo.toml
workspace: /Users/ksmadhukar/upstream-repos/kata-containers/Cargo.toml
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.47s
     Running unittests src/lib.rs (/Users/ksmadhukar/upstream-repos/kata-containers/target/debug/deps/kata_agent_policy-aef4ba498af320fc)

running 44 tests
test policy::tests::test_bundled_policy::case_2 ... ok
test policy::tests::test_bundled_policy::case_1 ... ok
test policy::tests::test_allow_failures_bypass ... ok
test policy::tests::test_bundled_policy::case_4 ... ok
test policy::tests::test_bundled_policy::case_3 ... ok
test policy::tests::test_allow_except_exec_denies_exec ... ok
test policy::tests::test_copyfile_translation::case_3 ... ok
test policy::tests::test_annotation_key_allowlisting ... ok
test policy::tests::test_copyfile_translation::case_1 ... ok
test policy::tests::test_copyfile_translation::case_2 ... ok
test policy::tests::test_copy_file_type_path_cases::case_3 ... ok
test policy::tests::test_copyfile_translation::case_4 ... ok
test policy::tests::test_copyfile_translation::case_5 ... ok
test policy::tests::test_copyfile_translation::case_6 ... ok
test policy::tests::test_copyfile_translation::case_7 ... ok
test policy::tests::test_copyfile_translation::case_8 ... ok
test policy::tests::test_copy_file_type_path_cases::case_8 ... ok
test policy::tests::test_copy_file_type_path_cases::case_4 ... ok
test policy::tests::test_copy_file_type_path_cases::case_1 ... ok
test policy::tests::test_copy_file_type_path_cases::case_5 ... ok
test policy::tests::test_copy_file_type_path_cases::case_7 ... ok
test policy::tests::test_create_sandbox_hook_path_blocked ... ok
test policy::tests::test_create_sandbox_kernel_modules_blocked ... ok
test policy::tests::test_copy_file_type_path_cases::case_6 ... ok
test policy::tests::test_create_sandbox_pidns_blocked ... ok
test policy::tests::test_copy_file_type_path_cases::case_2 ... ok
test policy::tests::test_create_sandbox_valid ... ok
test policy::tests::test_exec_process_selinux_blocked ... ok
test policy::tests::test_initialize_missing_file_errors ... ok
test policy::tests::test_exec_process_wrong_cmd ... ok
test policy::tests::test_exec_process_valid ... ok
test policy::tests::test_set_policy_invalid_inputs::case_1 ... ok
test policy::tests::test_prints_captured_on_allow_and_deny ... ok
test policy::tests::test_set_policy_invalid_inputs::case_2 ... ok
test policy::tests::test_copy_file_symlink_prints ... ok
test policy::tests::test_regex_sandbox_name_matching::case_1 ... ok
test policy::tests::test_regex_sandbox_name_matching::case_3 ... ok
test policy::tests::test_regex_sandbox_name_matching::case_2 ... ok
test policy::tests::test_initialize_loads_policy ... ok
test policy::tests::test_validate_and_invalid_input ... ok
test policy::tests::test_set_policy_replaces_previous_policy ... ok
test policy::tests::test_update_routes_request::case_3 ... ok
test policy::tests::test_update_routes_request::case_2 ... ok
test policy::tests::test_update_routes_request::case_1 ... ok

test result: ok. 44 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s
```

Assisted-by: IBM Bob

### CODE REVIEW

File: src/agent/policy/src/policy.rs [hunk 1]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+        let mut p = AgentPolicy::new();
+        p.set_policy(rego).await.unwrap();
+        p
+    }
+
+    /// Assert a single `allow_request` call on an already-loaded engine.
+    ///
+    /// - `want_allowed`   — expected `allowed` flag.
+    /// - `want_print`     — if `Some(s)`, asserts `prints.contains(s)`.
+    /// - `want_not_print` — if `Some(s)`, asserts `!prints.contains(s)`.
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:411]: This isn't a parameter of the function you are commenting?

Comment (stevenhorsman) [reply src/agent/policy/src/policy.rs:411]: Is this supposed to be on the next function?

File: src/agent/policy/src/policy.rs [hunk 2]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
-                    "test case {}\nunexpected success: {:?}",
-                    &test_case.name,
-                    output_res
-                )
+                    !want_allowed,
+                    "allow_request({ep}) returned Err ({e}), expected allow=true",
+                    ep = ep,
+                    e = e,
+                );
+                return; // Err counts as denied — nothing more to check
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:443]: Can we check the error matches what we expect?

File: src/agent/policy/src/policy.rs [hunk 3]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+            allowed,
+            want_allowed,
+            "ep={ep} input={input}: got allowed={allowed}, want {want_allowed}\nprints={prints}",
+            ep = ep,
+            input = input,
+            allowed = allowed,
+            want_allowed = want_allowed,
+            prints = prints,
+        );
+        if let Some(needle) = want_print {
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:456]: What is `needle` representing?

File: src/agent/policy/src/policy.rs [hunk 4]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+        }
+    }
+
+    // =========================================================================
+    // CopyFileRequest translation tests (pre-existing, ported to rstest)
+    // =========================================================================
+
+    // Successful conversions: (file_mode, path, data, expected_output)
+    #[rstest]
+    #[case(
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:482]: As with #13078, please use the rstest `case::<description>` approach to label tests rather than implementing your own.

File: src/agent/policy/src/policy.rs [hunk 5]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+        "/../../kata-opa/allow-set-policy.rego"
+    );
+
+    // =========================================================================
+    // Helpers
+    // =========================================================================
+
+    /// Load a bundled .rego file directly into a fresh engine, bypassing
+    /// `initialize()` so no log file or default-policy path is required.
+    async fn policy_with_file(path: &str) -> AgentPolicy {
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:394]: Why is this labelled as async, when it doesn't contain an await?

File: src/agent/policy/src/policy.rs [hunk 6]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+        let mut p = AgentPolicy::new();
+        p.engine.add_policy_from_file(path).unwrap();
+        p
+    }
+
+    /// Load an inline Rego string into a fresh engine via `set_policy()`.
+    async fn policy_with_str(rego: &str) -> AgentPolicy {
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:401]: Same here

File: src/agent/policy/src/policy.rs [hunk 7]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+        #[case] endpoint: &str,
+        #[case] expected: bool,
+    ) {
+        let mut p = policy_with_file(policy_file).await;
+        let req = match endpoint {
+            "SetPolicyRequest" => req_json!(SetPolicyRequest {
+                policy: "x".to_owned(),
+                ..Default::default()
+            }),
+            _ => req_json!(ExecProcessRequest::default()),
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:688]: Can you match against `ExecProccessRequest` separate and then just use the default for `panic!("unknown endpoint")` as there are many, many endpoints that don't match either, so it's fragile to just assume that non SetPolicy is ExecProcess?

File: src/agent/policy/src/policy.rs [hunk 8]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+            path: "/tmp/test".to_owned(),
+            ..Default::default()
+        });
+        let deny_req = req_json!(PolicyCopyFileRequest {
+            path: "/etc/passwd".to_owned(),
+            ..Default::default()
+        });
+
+        // Allow path: both prints fire — check for terminal print (implies opening also fired).
+        assert_allow_ext(
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:731]: This has None for `want_not_print`, so why isn't it calling `assert_allow`?

File: src/agent/policy/src/policy.rs [hunk 9]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+            "CopyFileRequest",
+            &allow_req,
+            true,
+            Some("CopyFileRequest: allowed"),
+            None,
+        )
+        .await;
+
+        // Deny path: opening print fires, terminal does not.
+        assert_allow_ext(
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:742]: If this is the only user of `want_not_print`, did you consider just moving this check out rather than having a dedicated helper, or are you anticipating tests extending to use it?

File: src/agent/policy/src/policy.rs [hunk 10]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+    // -------------------------------------------------------------------------
+    // CopyFileRequest file_type matrix — Regular / Directory / Symlink branches
+    // from rules.rego allow_copy_file, including traversal + symlink-target checks.
+    //
+    // A shared const carries the policy used by both the parametrized table and
+    // the dedicated print-assertion test, avoiding duplicate inline Rego strings.
+    // -------------------------------------------------------------------------
+
+    /// Shared CopyFileRequest policy used by test_copy_file_type_path_cases
+    /// and test_copy_file_symlink_prints.
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:762]: Is this comment just AI Slop? It's describing the callers not the content, so I'm not sure of the value?

File: src/agent/policy/src/policy.rs [hunk 11]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+            count(i_process.ApparmorProfile) == 0
+            print("ExecProcessRequest: Args =", i_process.Args)
+            i_process.Args[0] == "/bin/sh"
+            print("ExecProcessRequest: true")
+        }
+    "#;
+
+    /// Build an ExecProcessRequest with no string_user (serialises as null)
+    /// and the given process args + optional SelinuxLabel.
+    fn exec_req(args: &[&str], selinux_label: &str) -> ExecProcessRequest {
```
Comment (stevenhorsman) [inline src/agent/policy/src/policy.rs:990]: If selinux_label is optional, then why not make it `Option<&str>`?

**Review by stevenhorsman — COMMENTED** [review]
Thanks for trying to improve the coverage. Here are some comments to start with. I think this needs some attention as it just appears to mostly AI generated (without correct attribution in the commit message), leading to non-idiomatic code and comments that aren't consistent with the rest of our code and style.

**Comment by sprt** [comment]
@danmihai1 Seems redundant with the testing we already have?

**Comment by stevenhorsman** [comment]
> @danmihai1 Seems redundant with the testing we already have?

Hey, could you expand on this out of interest. Are you talking about genpolicy integration tests, or is the genpolicy component (which is a separate, optional tool [not part of a kata image] for generating the policy) unit testing functions in the agent (a different component that enforces the policy)? If so that seems weird/wrong, so I guess I'm missing something?

**Comment by danmihai1** [comment]
> > @danmihai1 Seems redundant with the testing we already have?
> 
> Hey, could you expand on this out of interest. Are you talking about genpolicy integration tests, or is the genpolicy component (which is a separate, optional tool [not part of a kata image] for generating the policy) unit testing functions in the agent (a different component that enforces the policy)? If so that seems weird/wrong, so I guess I'm missing something?

@SantoshMadhukar-K , thanks for offering help! 

@stevenhorsman, I guess @sprt was referring to these other tests, that also evaluate policy in a test process:

https://github.com/kata-containers/kata-containers/blob/3a9fd86ccc710ce21d477c888c400187889faaae/src/tools/genpolicy/tests/policy/main.rs#L177

It would be great if any of you folks could compare these two types of tests, and share the comparison results. Right now, I don't know if some of these test types are more useful than the others, or if keeping both types would be better.

Otherwise I will try to compare on my own when I can find some time.

CC:
@burgerdev , who designed src/tools/genpolicy/tests if I remember correctly.
@manuelh-dev , who helped with these tests recently.

File: src/agent/policy/src/policy.rs [hunk 12]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+                ep = ep,
+                needle = needle,
+                prints = prints,
+            );
+        }
+    }
+
+    // =========================================================================
+    // CopyFileRequest translation tests (pre-existing, ported to rstest)
+    // =========================================================================
```
Comment (burgerdev) [inline src/agent/policy/src/policy.rs:478]: Why not just keep the existing tests?

File: src/agent/policy/src/policy.rs [hunk 13]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+                result.is_err(),
+                "test case {_label}: expected Err, got {result:?}",
+                _label = _label,
+                result = result,
+            ),
+        }
+    }
+
+    // =========================================================================
+    // OPA engine tests
```
Comment (burgerdev) [inline src/agent/policy/src/policy.rs:549]: I don't think Kata should be testing regorus...

File: src/agent/policy/src/policy.rs [hunk 14]
```diff
@@ -354,141 +354,969 @@ mod tests {
...
+        };
+        assert_allow(&mut p, endpoint, &req, expected, None).await;
+    }
+
+    // =========================================================================
+    // Advanced tests — patterns from src/tools/genpolicy/rules.rego
+    //
+    // Each test uses an inline policy mirroring a specific rules.rego pattern
+    // and asserts both the allow/deny flag AND specific print() trace output.
+    // =========================================================================
```
Comment (burgerdev) [inline src/agent/policy/src/policy.rs:698]: I think this is not maintainable - we're not going to keep these tests in sync with `rules.rego`, and I also doubt whether these tests are contributing anything.

**Review by burgerdev — CHANGES_REQUESTED** [review]
Sorry, I don't think this PR is adding anything useful - added explanations to the individual subsections.

One alternative idea that we could pursue, between unit tests and integration tests, would be OPA unit tests. These could cover some of the trickier nested functions that we currently only exercise with full `CreateContainerRequest`s. I wrote an example here: https://github.com/edgelesssys/contrast/blob/5cac86a8cfe3258f6426177c8d3295da2a6ad011/packages/by-name/kata/source/rules_test.rego. It's not straightforward because OPA requires some of the `policy_data` members to be populated, but still better than no tests on that level.

**Comment by burgerdev** [comment]
> > @danmihai1 Seems redundant with the testing we already have?
> 
> Hey, could you expand on this out of interest. Are you talking about genpolicy integration tests, or is the genpolicy component (which is a separate, optional tool [not part of a kata image] for generating the policy) unit testing functions in the agent (a different component that enforces the policy)? If so that seems weird/wrong, so I guess I'm missing something?

The `src/agent/policy` crate only makes sense combined with `genpolicy`, and `genpolicy` Rust integration tests (not e2e!) use the crate directly (that's why it's not part of the core agent crate anymore). These will hopefully become proper unit tests (or at least hermetic) with #13702. Does that answer the question?

**Comment by stevenhorsman** [comment]
> The `src/agent/policy` crate only makes sense combined with `genpolicy`, and `genpolicy`

I don't want to have a big debate about this, as it's orthogonal to the usefulness of this PR, but I feel like I need to point out that I strongly disagree on this point - the policy pre-existed genpolicy, so you are saying that we should have blocked it as it didn't "make sense"? There are also multiple downstream products that document policy (as it is completely crucial to locking down confidential VM), but don't mandate how the policy is created, so I don't think it's appropriate for us to just merge the two together as the same feature.

**Comment by danmihai1** [comment]
> > The `src/agent/policy` crate only makes sense combined with `genpolicy`, and `genpolicy`
> 
> I don't want to have a big debate about this, as it's orthogonal to the usefulness of this PR, but I feel like I need to point out that I strongly disagree on this point - the policy pre-existed genpolicy, so you are saying that we should have blocked it as it didn't "make sense"? There are also multiple downstream products that document policy (as it is completely crucial to locking down confidential VM), but don't mandate how the policy is created, so I don't think it's appropriate for us to just merge the two together as the same feature.

I agree with @stevenhorsman here: we should do our best to not break non-genpolicy Policies. I guess we might break them every once in a while because we had to make changes for some important reasons, but in general other Policies should continue to work.

I have just read more carefully through this PR's diff. Based on that reading, **I vote to accept these tests** (perhaps after improving them a bit if needed, as we do with for any other PR). 

I like how easy it is to add a new Policy text to these tests, and to construct input Requests having more unusual field values.

Some of these test cases are validating rule patterns used by genpolicy's rules.rego. If we'll accept these tests, maintainers will have to judge in the future if updating rules.rego patterns should result in an Agent unit test update too. It's possible that in some cases the Agent unit tests will continue to test old/outdated patterns, but that's not a big deal in my opinion.

@manuelh-dev , @Redent0r , any thoughts on these new tests? 
@burgerdev , did Steve and I change your earlier opinion at all?
Any other maintainers want to chime in?

Thanks folks!

**Comment by stevenhorsman** [comment]
Thanks for the comments Dan - I completely agree that we need to work out the right boundaries and it could be that these are "overfit" to our current rules and should be more abstract to check the policy enforcement here, including the edge cases and worry about rule complexity elsewhere. I'm also open to the idea that this are not needed at all and covered elsewhere, or we should migrate/adopt some genpolicy tests that provide better coverage here. I think Santosh just felt there was a coverage gap (and wasn't aware of the genpolicy tests that may fill it) and wanted to help improve that, so if the community and particularly SMEs of policy felt this is wrong then we are happy to close it completely. Thanks for the good discussion all!

**Comment by burgerdev** [comment]
As I mentioned in the comments on the individual blocks, I don't see the added value, but maybe that could come later? The ability to load a policy document and execute it on RPCs is sufficiently tested, imho. That being said, I don't mind merging these, as they are probably also low maintenance effort.

Regarding policy use without `genpolicy`: maybe I got the history wrong, but I thought `genpolicy` was the target from the original design: https://github.com/kata-containers/kata-containers/issues/7573. I'm sceptical of using other policies, because it's already hard enough to keep the `genpolicy` one secure. Suggestions I've seen in the wild are usually very insecure (e.g. ["restrictive policy"](https://confidentialcontainers.org/docs/features/initdata/#example-restrictive-agent-policy)), and I think telling users to author their own policies is not responsible.

File: src/agent/policy/src/policy.rs (hunk 13 above)
Comment (Redent0r) [reply src/agent/policy/src/policy.rs:549]: true, I think this is the job of https://github.com/microsoft/regorus/tree/main/tests

**Comment by Redent0r** [comment]
> As I mentioned in the comments on the individual blocks, I don't see the added value, but maybe that could come later? The ability to load a policy document and execute it on RPCs is sufficiently tested, imho. That being said, I don't mind merging these, as they are probably also low maintenance effort.
> 
> Regarding policy use without `genpolicy`: maybe I got the history wrong, but I thought `genpolicy` was the target from the original design: #7573. I'm sceptical of using other policies, because it's already hard enough to keep the `genpolicy` one secure. Suggestions I've seen in the wild are usually very insecure (e.g. ["restrictive policy"](https://confidentialcontainers.org/docs/features/initdata/#example-restrictive-agent-policy)), and I think telling users to author their own policies is not responsible.

+1 here. Im not sure what these tests are adding coverage for (what change would break them). I feel they are mostly making sure the OPA engine (regorus) evaluates (accepts or rejects) as expected, given a request and a policy text. If so, I think that coverage should go in https://github.com/microsoft/regorus/tree/main/tests

**Comment by danmihai1** [comment]
OK, it looks like the majority was different from my vote. So, unless those conditions will change, I give up. Thanks, everyone!

**Comment by stevenhorsman** [comment]
OK, I think based on the comments from maintainers then we should just close this PR. Thanks for the feedback

**Comment by SantoshMadhukar-K** [comment]
Thanks (@stevenhorsman and maintainers) for the reviews and feedback .

Going through the feedback and existing test suite in `src/tools/genpolicy/tests/policy/main.rs`, I agree that these agent policy code is already well-covered by integration tests using generated policies.

Closing this PR. Thanks again for your time and guidance!
